pkg/controllers/user/multiclusterapp/controller.go

```diff
@@ -0,0 +1,267 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	globalScopeAnswersKey     = "global"
+	creatorIDAnn              = "field.cattle.io/creatorId"
+	multiClusterAppIDSelector = "mClusterApp"
+	projectIDFieldLabel       = "field.cattle.io/projectId"
+	multiClusterAppPrefix     = "mcapp-"
+)
+
+type MCAppController struct {
+	AppClient             pv3.AppInterface
+	AppLister             pv3.AppLister
+	MultiClusterAppClient v3.MultiClusterAppInterface
+	MultiClusterAppLister v3.MultiClusterAppLister
+	NsClient              corev1.NamespaceInterface
+	NsLister              corev1.NamespaceLister
+	TemplateVersionClient v3.TemplateVersionInterface
+	ClusterLister         v3.ClusterLister
+	ProjectLister         v3.ProjectLister
+	ClusterName           string
+}
+
+func Register(ctx context.Context, cluster *config.UserContext) {
+	m := &MCAppController{
+		AppClient:             cluster.Management.Project.Apps(""),
+		AppLister:             cluster.Management.Project.Apps("").Controller().Lister(),
+		NsClient:              cluster.Core.Namespaces(""),
+		NsLister:              cluster.Core.Namespaces("").Controller().Lister(),
+		MultiClusterAppClient: cluster.Management.Management.MultiClusterApps(""),
+		MultiClusterAppLister: cluster.Management.Management.MultiClusterApps("").Controller().Lister(),
+		TemplateVersionClient: cluster.Management.Management.TemplateVersions(""),
+		ClusterLister:         cluster.Management.Management.Clusters("").Controller().Lister(),
+		ProjectLister:         cluster.Management.Management.Projects("").Controller().Lister(),
+		ClusterName:           cluster.ClusterName,
+	}
+	m.MultiClusterAppClient.AddHandler(ctx, "multi-cluster-app-controller", m.sync)
+}
+
+func (m *MCAppController) sync(key string, mcapp *v3.MultiClusterApp) (runtime.Object, error) {
+	if mcapp == nil || mcapp.DeletionTimestamp != nil {
+		_, mcappName, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			return mcapp, err
+		}
+		return m.deleteApps(mcappName)
+	}
+
+	metaAccessor, err := meta.Accessor(mcapp)
+	if err != nil {
+		return mcapp, err
+	}
+	creatorID, ok := metaAccessor.GetAnnotations()[creatorIDAnn]
+	if !ok {
+		logrus.Errorf("MultiClusterApp %v has no creatorId annotation. Cannot create apps for %v", metaAccessor.GetName(), mcapp.Name)
+		return mcapp, nil
+	}
+	ann := make(map[string]string)
+	ann[creatorIDAnn] = creatorID
+	set := labels.Set(map[string]string{multiClusterAppIDSelector: mcapp.Name})
+
+	answerMap, answerFound := m.createAnswerMap(mcapp.Spec.Answers)
+	// create app object spec
+	// create the externalID field, it's also present on the templateVersion. So get the templateVersion and read its externalID field
+	tv, err := m.TemplateVersionClient.Get(mcapp.Spec.TemplateVersionName, metav1.GetOptions{})
```

Comment:
any reason we are reading it from the client, not from cache?

Reply:
No reason, but it's only happening once per multiclusterapp launch, but I'll change to listers